sdks/java/io/iceberg/src/main/java/org/apache/beam/io/iceberg/CombinedScanReader.java

```diff
@@ -0,0 +1,214 @@
+import java.util.ArrayDeque;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.values.Row;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.data.avro.DataReader;
+import org.apache.iceberg.data.orc.GenericOrcReader;
+import org.apache.iceberg.data.parquet.GenericParquetReaders;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.encryption.InputFilesDecryptor;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.orc.ORC;
+import org.apache.iceberg.parquet.Parquet;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CombinedScanReader extends BoundedSource.BoundedReader<Row> {
+  private static final Logger LOG = LoggerFactory.getLogger(CombinedScanReader.class);
+
+  private final IcebergBoundedSource source;
+  private final Table table;
+
+  private final @Nullable CombinedScanTask task;
+
+  private final Schema schema;
+
+  transient @Nullable org.apache.iceberg.Schema project;
+  transient @Nullable FileIO io;
+  transient @Nullable InputFilesDecryptor decryptor;
+  transient @Nullable Queue<FileScanTask> files;
+  transient @Nullable CloseableIterator<Record> baseIter;
+  transient @Nullable Record current;
+
+  public CombinedScanReader(
+      IcebergBoundedSource source, @Nullable CombinedScanTask task, Schema schema) {
+    this.source = source;
+    this.table =
+        checkStateNotNull(
+            source.table(), "CombinedScanReader requires an IcebergBoundedSource with a table");
+    this.task = task;
+    this.schema = schema;
+    if (this.schema != null) {
+      project = SchemaHelper.convert(this.schema);
+    }
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    if (task == null) {
+      return false;
+    }
+
+    EncryptionManager encryptionManager = table.encryption();
+    io = table.io();
+    decryptor = new InputFilesDecryptor(task, io, encryptionManager);
+
+    files = new ArrayDeque<>();
+    files.addAll(task.files());
+
+    return advance();
+  }
+
+  @Override
+  public boolean advance() throws IOException {
+    Queue<FileScanTask> files =
+        checkStateNotNull(this.files, "files null in advance() - did you call start()?");
+    InputFilesDecryptor decryptor =
+        checkStateNotNull(this.decryptor, "decryptor null in adance() - did you call start()?");
+
+    // This is a lie, but the most expedient way to work with IcebergIO's
```

Comment:
Update/remove comment ?

Reply:
Actually I added this comment to explain why the type was fake. I reworded it.